spotbugsTestCases/src/java/overridableMethodCall/DirectReadObjectStreamMethods.java

```diff
@@ -0,0 +1,12 @@
+package overridableMethodCall;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+
+public class DirectReadObjectStreamMethods {
+
+    private void readObject(final ObjectInputStream stream) throws IOException, ClassNotFoundException {
```

Comment:
Checking the exceptional methods indirectly doesn't work due to a limitation of the detector.

```java
else if (item.getRegisterNumber() == 0
                    && (getXMethod().isPrivate() || getXMethod().isFinal()))
```

In my understanding we only check indirect calls, if they are invoked on `this`. I think it's incorrect, and we should check every other cases, but I'm afraid if I change this part weird things will start happening with the detector. I can however add an exception here for `readObject()`.

Reply:
I tried it, the checking of the exceptional methods indirectly does work, there is no FP reported when these methods are called indirectly. Can you please add the testcase for it?

Reply:
I added a testcase.

> I tried it, the checking of the exceptional methods indirectly does work, there is no FP reported when these methods are called indirectly.

You see no false positives, because that case is ignored by the check completely. It's a coincidence that for these specific methods this is the correct behaviour.

However this limitation does produce false negatives. Take a look at this example:
```java
private void readObject(final ObjectInputStream stream) throws IOException {
    stream.close();
}
```
```
> spotbugs ...
M V MC: Overridable method close is called from readObject.  At Test.java:[line 6]
```

```java
private void indirect(final ObjectInputStream stream) throws IOException {
    stream.close();
}
private void readObject(final ObjectInputStream stream) throws IOException {
    indirect(stream);
}
```
```
> spotbugs ...
```

Reply:
Thank you for the explanation 😄 
It would be great if you could take a look into this issue. Feel free to modify the code, this is the reason we have the tests for, the tests should show it, if something breaks. If not, even adding this test, with commenting that this is an FP, the limitation of the detector would also be really helpful. However, IMO it's better to have false negatives, than having a really noisy detector, which the users ignore and disable.

Reply:
I was actually experimenting with removing this limitation, but the result was too noisy for me.

```java
private void indirect() {
    System.out.println("foo");
}
private void readObject(final ObjectInputStream stream) {
    indirect();
}
```
`println()` is also overridable and will be reported as an error. Also if I understand it correctly, for `println()` we don't care if it's actually overridden or not, because even if it's overrideen it can't access the current object's state, right?

Reply:
Yes, the override of `System.out.println()` is not what this detector is about. Even if that's overriden, that won't be the one called.
Thank you for taking a look at the problem.